controllers/catalogsource_controller.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/operator-framework/operator-controller/internal/resolution/entity_sources/catalogsource"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	defaultCatalogSourceSyncInterval     = 5 * time.Minute
+	defaultRegistryGRPCConnectionTimeout = 10 * time.Second
+
+	eventTypeNormal  = "Normal"
+	eventTypeWarning = "Warning"
+
+	eventReasonCacheUpdated      = "BundleCacheUpdated"
+	eventReasonCacheUpdateFailed = "BundleCacheUpdateFailed"
+)
+
+type CatalogSourceReconcilerOption func(reconciler *CatalogSourceReconciler)
+
+func WithRegistryClient(registry catalogsource.RegistryClient) CatalogSourceReconcilerOption {
+	return func(reconciler *CatalogSourceReconciler) {
+		reconciler.registry = registry
+	}
+}
+
+// applyDefaults applies default values to empty CatalogSourceReconciler fields _after_ options have been applied
+func applyDefaults() CatalogSourceReconcilerOption {
+	return func(reconciler *CatalogSourceReconciler) {
+		if reconciler.registry == nil {
+			reconciler.registry = catalogsource.NewRegistryGRPCClient(defaultRegistryGRPCConnectionTimeout)
+		}
+	}
+}
+
+type CatalogSourceReconciler struct {
+	sync.RWMutex
+	client.Client
+	scheme   *runtime.Scheme
+	registry catalogsource.RegistryClient
+	recorder record.EventRecorder
+	cache    map[string]map[deppy.Identifier]*input.Entity
+}
+
+func NewCatalogSourceReconciler(client client.Client, scheme *runtime.Scheme, recorder record.EventRecorder, options ...CatalogSourceReconcilerOption) *CatalogSourceReconciler {
+	reconciler := &CatalogSourceReconciler{
+		RWMutex:  sync.RWMutex{},
+		Client:   client,
+		scheme:   scheme,
+		recorder: recorder,
+		cache:    map[string]map[deppy.Identifier]*input.Entity{},
+	}
+	// apply options
+	options = append(options, applyDefaults())
+	for _, option := range options {
+		option(reconciler)
+	}
+
+	return reconciler
+}
+
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=catalogsources,verbs=get;list;watch
+
+func (r *CatalogSourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithName("catalogsource-controller")
+	l.V(1).Info("starting")
+	defer l.V(1).Info("ending")
+
+	var catalogSource = &v1alpha1.CatalogSource{}
+	if err := r.Client.Get(ctx, req.NamespacedName, catalogSource); err != nil {
+		if errors.IsNotFound(err) {
+			r.dropSource(req.String())
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	entities, err := r.registry.ListEntities(ctx, catalogSource)
+	if err != nil {
+		r.recorder.Event(catalogSource, eventTypeWarning, eventReasonCacheUpdateFailed, fmt.Sprintf("Failed to update bundle cache from %s/%s: %v", catalogSource.GetNamespace(), catalogSource.GetName(), err))
+		return ctrl.Result{Requeue: true}, err
```

Comment:
For later: I wonder if there's a way to propagate this to the Operator object.

Something along the lines of:
1. mark this cache as stale in a way that the operator controller can see it.
2. when reconciling an Operator, if any of the caches are stale, set a separate condition (e.g. `UsingStaleCatalog` or `CatalogUnavailable` or something) 

We could even later build on that and have a knob in the Operator spec that gives the cluster admin some control about how the resolver should handle that situation. In OLMv0 (iirc), we fail resolution if any of the catalog sources are unavailable, and I'm not sure there's a way to say "ok yeah, just resolve without that catalog"

Reply:
I'm not sure we want to fail outright for all resolution for a single stale/unreachable catalogSource. Currently the catalog reconciler should continue resolving with whatever the last healthy state of the failed catalog was despite emitting events about this failure.

It may be worth considering some way to mark different levels of staleness on the caches, like the number of updates that have failed due to an unreachable catalogSource and to have that as a guide to determine how to deal with the cached version.

Reply:
I'm not sure we want to fail outright for all resolution for a single stale/unreachable catalogSource. Currently the catalog reconciler should continue resolving with whatever the last healthy state of the failed catalog was despite emitting events about this failure.

:+1: Yeah agreed. In my mind we can evolve this over time. Something along the lines of:
- Iteration 1: we resolve with whatever we have, even if expected stuff is missing or stale
- Iteration 2: we add detection for unreachable catsrcs and tell the user that resolution is using incomplete data. (My initial instinct would be to never use a stale cache. If a catsrc worked at one point, we populated the cache, and then it becomes unreachable, I'm thinking we just empty out the cache so that we don't use old data.)
- Iteration 3: we give the user a knob that let's them decide whether resolution should fail on unreachable catsrcs.

Reply:
In the statement "continue resolving with stale cache", what does "continue resolving" mean? Does it install dependencies, that admins will have to clean up if that wasn't the intended dependency resolution from the user's perspective? (think it installed etcd operator from the community-operator index but I absolutely cannot have any workload from the community-operator index running, so now I have to re-trigger the resolution once the certified-operators catalog is available again, and for that I need to uninstall a bunch of dependencies)

Reply:
> it installed etcd operator from the community-operator index but I absolutely cannot have any workload from the community-operator index running, so now I have to re-trigger the resolution once the certified-operators catalog is available again, and for that I need to uninstall a bunch of dependencies

I'm imagining you're talking about a scenario where both catalogSources are installed at once, but the certified-operator index takes some time to become available, during which one round of resolution completes with the community-operator index. In that case, we'd certainly run into the problem of having incorrect operators installed on cluster. We might be able to work around this by extending the operator object to have more constraints over what gets selected - like maybe a restriction on a custom property like the publisher or similar. 

What I mean by "continue resolving with stale cache" is more like if we have a catalog v-1 on cluster which has its underlying source updated to serve v-2. If the caching fails because the catalog source is unreachable or similar, the resolver would continue using its stale v-1 cache. If this isn't noticed or fixed in time, we could have dependencies installed that can be difficult to clean up. 

One solution is to void any cached catalogSource that isn't available for even one sync - so we'd need to rebuild the cache any time a catalogSource becomes temporarily unavailable. This might have some impact on the actual cache performance. Another is to have cache being marked stale, stopping any attempt at resolution that doesn't specify it's ok to resolve from a stale cache.